### PR TITLE
Optimize renderer base

### PR DESCRIPTION
**Pull request by Page-**
Connects-to: # 
See: <url> 
Depends-on: <url> 
Change-type: major|minor|patch 

---
##### Contributor checklist

- [ ] I have regenerated screenshots for any affected components with `npm run generate-screenshots`
- [ ] I have regenerated jest snapshots for any affected components with `npm run jest -- -u`

##### Reviewer Guidelines
- When submitting a review, please pick:
  - '*Approve*' if this change would be acceptable in the codebase (even if there are minor or cosmetic tweaks that could be improved).
  - '*Request Changes*' if this change would not be acceptable in our codebase (e.g. bugs, changes that will make development harder in future, security/performance issues, etc).
  - '*Comment*' if you don't feel you have enough information to decide either way (e.g. if you have major questions, or you don't understand the context of the change sufficiently to fully review yourself, but want to make a comment)
---

### CODE REVIEW

File: src/components/Renderer/index.tsx [hunk 1]
```diff
@@ -25,14 +25,13 @@ export const getValue = (
...
 	// Fall back to schema's default value if value is undefined
-	return calculatedValue !== undefined
-		? calculatedValue
-		: get(schema, 'default');
+	return calculatedValue ?? schema?.default;
 };
 
-const widgetWrapperUiOptionKeys = keys(WidgetWrapperUiOptions);
+// Convert these to lodash paths once here to avoid having to do it for every single render
+const widgetWrapperUiOptionKeyPaths = keys(WidgetWrapperUiOptions).map(toPath);
```
Comment (Page-) [inline src/components/Renderer/index.tsx:34]: This preparsing is the main win, the path conversion in general is pretty quick but when rendering a large list of items it gets called enough times to be noticeable, and also allocates arrays each time which will contribute to time spent in the gc

File: src/components/Renderer/index.tsx [hunk 2]
```diff
@@ -25,14 +25,13 @@ export const getValue = (
 	schema?: JSONSchema,
 	uiSchema?: UiSchema,
 ) => {
-	const calculatedValue = get(uiSchema, 'ui:value', value);
+	const calculatedValue = uiSchema?.['ui:value'] ?? value;
```
Comment (Page-) [inline src/components/Renderer/index.tsx:28]: These `get` changes were primarily to clean up/simplify the code so it was easier to see what was going on once it has all been transpiled/etc and it does also improve the ability for typescript to apply type-checking

File: src/components/Renderer/index.tsx [hunk 3]
```diff
@@ -25,14 +25,13 @@ export const getValue = (
...
 	// Fall back to schema's default value if value is undefined
-	return calculatedValue !== undefined
-		? calculatedValue
-		: get(schema, 'default');
+	return calculatedValue ?? schema?.default;
```
Comment (thgreasi) [inline src/components/Renderer/index.tsx:30]: From the comment it looks like the `undefined` only check might have been intentional :thinking: 
@grahammcculloch ^ ?

**Review by thgreasi — COMMENTED** [review]
I'm not confident about the `undefined ? :` => `??` change, but other than that LGTM :+1:

File: src/components/Renderer/index.tsx (hunk 3 above)
Comment (grahammcculloch) [reply src/components/Renderer/index.tsx:30]: @thgreasi yes exactly. `null` is a valid value in this case so we want to explicitly test for `undefined`.

**Review by grahammcculloch — CHANGES_REQUESTED** [review]
Mostly looks good - just one issue where we still need to explicitly check for `undefined`.

Thanks for these optimizations @Page- !

File: src/components/Renderer/index.tsx (hunk 3 above)
Comment (Page-) [reply src/components/Renderer/index.tsx]: Urg, you're right. I saw so many cases around the codebase where it's pre `??` that I assumed this was one as well. I've fixed it so explicit `null` vs `undefined` in `ui:value` will also be respected

**Review by grahammcculloch — APPROVED** [review]
👍

**Pull request merged**